compiler/rustc_const_eval/src/interpret/discriminant.rs

```diff
@@ -241,10 +241,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         variant_index: VariantIdx,
     ) -> InterpResult<'tcx, Option<(ScalarInt, usize)>> {
         match self.layout_of(ty)?.variants {
-            abi::Variants::Single { index } => {
-                assert_eq!(index, variant_index);
```

Comment:
Removing this assertion was recommended by @RalfJung in https://github.com/rust-lang/rust/issues/125811#issuecomment-2158835857, and I agree with his assessment that it isn't a load-bearing assertion. An index mismatch here doesn't suggest a bug in producing the tag.

That said, removing this assertion is irrelevant to fixing the bug — the salient changes are all in `compiler/rustc_transmute/src/layout/tree.rs`. Those changes fix the ICE even with this assertion intact.

Reply:
I would suggest to replace the assertion by a comment, like
```
// The tag of a `Single` enum is like the tag of the niched variant: there's no tag
// as the discriminant is encoded entirely implicitly.
// If `write_discriminant` ever hits this case, we do a "validation read" to ensure
// the the right discriminant is encoded implicitly, so any attempt to write the
// wrong discriminant for a `Single` enum will reliably result in UB.
```